File: src/Features/ExcelUpload/UploadExcel.ts
import { z } from "zod";
import { Result } from "@/Common/Result";
import { ExcelErrors } from "@/Common/Errors";
import { writeFile, mkdir } from "fs/promises";
import { join } from "path";
import { randomUUID } from "crypto";
import { IFileStorage as ContainerIFileStorage } from "@/Infrastructure/DependencyInjection/Container";
import { IFileRepository } from "@/Common/Repositories/IFileRepository";

// Request Schema
export const UploadExcelRequestSchema = z.object({
  file: z.instanceof(File),
  userId: z.string().min(1, "사용자 ID가 필요합니다."),
});

export type UploadExcelRequest = z.infer<typeof UploadExcelRequestSchema>;

// Response Type
export interface UploadExcelResponse {
  fileId: string;
  fileName: string;
  fileSize: number;
  status: string;
  uploadedAt: Date;
}

// Validator
export class UploadExcelValidator {
  private static readonly validContentTypes = [
    "application/vnd.ms-excel",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "text/csv",
  ];

  private static readonly maxFileSize = 50 * 1024 * 1024; // 50MB

  static validate(request: UploadExcelRequest): Result<void> {
    const { file } = request;

    if (!file) {
      return Result.failure(ExcelErrors.EmptyFile);
    }

    if (!this.validContentTypes.includes(file.type)) {
      return Result.failure(ExcelErrors.InvalidFormat);
    }

    if (file.size > this.maxFileSize) {
      return Result.failure(ExcelErrors.TooLarge);
    }

    return Result.success(undefined);
  }
}

// File Storage Interface
export interface IFileStorage {
  uploadAsync(
    file: File,
    fileName: string
  ): Promise<Result<string>>;
}

// Local File Storage Implementation (for development)
export class LocalFileStorage implements IFileStorage {
  private uploadDir = join(process.cwd(), "uploads");

  async uploadAsync(
    file: File,
    fileName: string
  ): Promise<Result<string>> {
    try {
      // Create uploads directory if it doesn't exist
      await mkdir(this.uploadDir, { recursive: true });

      // Convert file to buffer
      const bytes = await file.arrayBuffer();
      const buffer = Buffer.from(bytes);

      // Write file
      const filePath = join(this.uploadDir, fileName);
      await writeFile(filePath, buffer);

      return Result.success(`/uploads/${fileName}`);
    } catch (error) {
      console.error("File storage error:", error);
      return Result.failure(ExcelErrors.ProcessingFailed);
    }
  }
}

// Handler
export class UploadExcelHandler {
  constructor(
    private fileRepository: IFileRepository,
    private fileStorage?: ContainerIFileStorage | IFileStorage
  ) {
    if (!this.fileStorage) {
      this.fileStorage = new LocalFileStorage();
    }
  }

  async handle(
    request: UploadExcelRequest
  ): Promise<Result<UploadExcelResponse>> {
    try {
      // Validate request
      const validationResult = UploadExcelValidator.validate(request);
      if (!validationResult.isSuccess) {
        return Result.failure(validationResult.error);
      }

      const { file, userId } = request;

      // Generate unique file identifier
      const fileExtension = file.name.split(".").pop();
      const fileName = `${randomUUID()}.${fileExtension}`;

      // Upload file to storage
      let uploadResult: Result<string>;
      
      if ('uploadAsync' in this.fileStorage && typeof this.fileStorage.uploadAsync === 'function') {
        uploadResult = await this.fileStorage.uploadAsync(file, fileName);
      } else {
        // Fallback for Container IFileStorage
        const buffer = Buffer.from(await file.arrayBuffer());
        const filePath = await (this.fileStorage as ContainerIFileStorage).save(buffer, fileName);
        uploadResult = Result.success(filePath);
      }
      
      if (!uploadResult.isSuccess) {
        return Result.failure(uploadResult.error);
      }

      // Save file metadata to database using repository
      const saveResult = await this.fileRepository.save({
        userId,
        fileName,
        originalName: file.name,
        fileSize: file.size,
        mimeType: file.type,
        uploadUrl: uploadResult.value,
        status: "PENDING",
      });

      if (!saveResult.isSuccess) {
        return Result.failure(saveResult.error);
      }

      const response: UploadExcelResponse = {
        fileId: saveResult.value,
        fileName: file.name,
        fileSize: file.size,
        status: "PENDING",
        uploadedAt: new Date(),
      };

      return Result.success(response);
    } catch (error) {
      console.error("Upload Excel handler error:", error);
      return Result.failure(ExcelErrors.ProcessingFailed);
    }
  }
}

File: src/Features/ExcelUpload/UploadExcel.cs
using ExcelApp.Common;
using FluentValidation;
using MediatR;

namespace ExcelApp.Features.ExcelUpload;

public static class UploadExcel
{
    // Request
    public record Request(
        Stream ExcelFile,
        string FileName,
        string ContentType,
        long FileSize,
        string UserId
    ) : IRequest<Result<Response>>;

    // Response
    public record Response(
        string FileId,
        string FileName,
        long FileSize,
        string Status,
        DateTime UploadedAt
    );

    // Validator
    public class Validator : AbstractValidator<Request>
    {
        private static readonly string[] ValidContentTypes = 
        {
            "application/vnd.ms-excel",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            "text/csv"
        };

        private const long MaxFileSize = 50 * 1024 * 1024; // 50MB

        public Validator()
        {
            RuleFor(x => x.FileName)
                .NotEmpty().WithMessage("파일명이 필요합니다.");

            RuleFor(x => x.ContentType)
                .Must(ct => ValidContentTypes.Contains(ct))
                .WithMessage("지원하지 않는 파일 형식입니다.");

            RuleFor(x => x.FileSize)
                .LessThanOrEqualTo(MaxFileSize)
                .WithMessage("파일 크기는 50MB를 초과할 수 없습니다.");

            RuleFor(x => x.ExcelFile)
                .NotNull().WithMessage("파일이 필요합니다.")
                .Must(stream => stream.CanRead)
                .WithMessage("파일을 읽을 수 없습니다.");

            RuleFor(x => x.UserId)
                .NotEmpty().WithMessage("사용자 ID가 필요합니다.");
        }
    }

    // Handler
    public class Handler : IRequestHandler<Request, Result<Response>>
    {
        private readonly IFileStorage _fileStorage;
        private readonly IFileRepository _fileRepository;
        private readonly ILogger<Handler> _logger;

        public Handler(
            IFileStorage fileStorage,
            IFileRepository fileRepository,
            ILogger<Handler> logger)
        {
            _fileStorage = fileStorage;
            _fileRepository = fileRepository;
            _logger = logger;
        }

        public async Task<Result<Response>> Handle(Request request, CancellationToken cancellationToken)
        {
            try
            {
                // Generate unique file identifier
                var fileId = Guid.NewGuid().ToString();
                var fileExtension = Path.GetExtension(request.FileName);
                var storedFileName = $"{fileId}{fileExtension}";

                // Upload file to storage
                var uploadResult = await _fileStorage.UploadAsync(
                    request.ExcelFile, 
                    storedFileName,
                    cancellationToken);

                if (!uploadResult.IsSuccess)
                {
                    _logger.LogError("Failed to upload file: {Error}", uploadResult.Error);
                    return Result.Failure<Response>(ExcelErrors.ProcessingFailed);
                }

                // Save file metadata to database
                var fileEntity = new FileEntity
                {
                    Id = fileId,
                    UserId = request.UserId,
                    FileName = storedFileName,
                    OriginalName = request.FileName,
                    FileSize = request.FileSize,
                    MimeType = request.ContentType,
                    UploadUrl = uploadResult.Value,
                    Status = FileStatus.Pending,
                    UploadedAt = DateTime.UtcNow
                };

                await _fileRepository.AddAsync(fileEntity, cancellationToken);

                var response = new Response(
                    fileEntity.Id,
                    fileEntity.OriginalName,
                    fileEntity.FileSize,
                    fileEntity.Status.ToString(),
                    fileEntity.UploadedAt
                );

                _logger.LogInformation(
                    "File uploaded successfully: {FileId} for user {UserId}", 
                    fileId, 
                    request.UserId);

                return Result.Success(response);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error during file upload");
                return Result.Failure<Response>(ExcelErrors.ProcessingFailed);
            }
        }
    }

    // Domain-specific interfaces (will be implemented in Infrastructure)
    public interface IFileStorage
    {
        Task<Result<string>> UploadAsync(Stream file, string fileName, CancellationToken cancellationToken);
    }

    public interface IFileRepository
    {
        Task AddAsync(FileEntity entity, CancellationToken cancellationToken);
    }

    // Domain entity
    public class FileEntity
    {
        public string Id { get; set; }
        public string UserId { get; set; }
        public string FileName { get; set; }
        public string OriginalName { get; set; }
        public long FileSize { get; set; }
        public string MimeType { get; set; }
        public string UploadUrl { get; set; }
        public FileStatus Status { get; set; }
        public DateTime UploadedAt { get; set; }
    }

    public enum FileStatus
    {
        Pending,
        Processing,
        Completed,
        Failed
    }
}